shared/rSharp.cpp

```diff
@@ -27,53 +27,99 @@ typedef intptr_t(CORECLR_DELEGATE_CALLTYPE* CallFullTypeNameDelegate)(RSharpGene
-
-	// STEP 1: Load HostFxr and get exported hosting functions
-	if (!load_hostfxr())
+	int returnValue = 0;
+	try
 	{
-		assert(false && "Failure: load_hostfxr()");
-		return;
-	}
+		args = CDR(args);
+
+		const SEXP element = CAR(args);
+		const char* libraryPath = CHAR(STRING_ELT(element, 0));
+
+		// if already loaded in this process, do not load again
+		if (load_assembly_and_get_function_pointer != nullptr)
+		{
+			returnValue = 0;
+			return make_int_sexp(1, &returnValue);
+		}
+
+		// STEP 1: Load HostFxr and get exported hosting functions
+		if (!load_hostfxr())
+		{
+			assert(false && "Failure: load_hostfxr()");
+			{
+				returnValue = 2;
+				return make_int_sexp(1, &returnValue);
+			}
+		}
 
 #ifdef WINDOWS
-	// STEP 2: Initialize and start the .NET Core runtime
-	size_t lengthInWideFormat = 0;
-	//Gets the length of libPath in terms of a wide string.
-	mbstowcs_s(&lengthInWideFormat, nullptr, 0, *libraryPath, 0);
-	char_t* wideStringLibraryPath = new char_t[lengthInWideFormat + 1];
-	//Copies the libraryPath to a wchar_t with the size lengthInWideFormat
-	mbstowcs_s(nullptr, wideStringLibraryPath, lengthInWideFormat + 1, *libraryPath, lengthInWideFormat);
+		// STEP 2: Initialize and start the .NET Core runtime
+		size_t lengthInWideFormat = 0;
+		//Gets the length of libPath in terms of a wide string.
+		mbstowcs_s(&lengthInWideFormat, nullptr, 0, libraryPath, 0);
+		char_t* wideStringLibraryPath = new char_t[lengthInWideFormat + 1];
+		//Copies the libraryPath to a wchar_t with the size lengthInWideFormat
+		mbstowcs_s(nullptr, wideStringLibraryPath, lengthInWideFormat + 1, libraryPath, lengthInWideFormat);
 #else
-	char_t* wideStringLibraryPath = *libraryPath;
+		char_t* wideStringLibraryPath = *libraryPath;
 #endif
 
-	char_t* wideStringPath = MergeLibraryPath(wideStringLibraryPath, STR("/RSharp.runtimeconfig.json"));
-	dotnetlib_path = MergeLibraryPath(wideStringLibraryPath, STR("/ClrFacade.dll"));
+		char_t* wideStringPath = MergeLibraryPath(wideStringLibraryPath, STR("/RSharp.runtimeconfig.json"));
+		dotnetlib_path = MergeLibraryPath(wideStringLibraryPath, STR("/ClrFacade.dll"));
 
-	load_assembly_and_get_function_pointer = nullptr;
-	load_assembly_and_get_function_pointer = get_dotnet_load_assembly(wideStringPath);
-	assert(load_assembly_and_get_function_pointer != nullptr && "Failure: get_dotnet_load_assembly()");
+		load_assembly_and_get_function_pointer = nullptr;
+		load_assembly_and_get_function_pointer = get_dotnet_load_assembly(wideStringPath);
+		
+		if(load_assembly_and_get_function_pointer == nullptr)
+		{
+			assert("Failure: get_dotnet_load_assembly()");
+			throw std::runtime_error("Failure: get_dotnet_load_assembly()");
+			returnValue = 3;
+			return make_int_sexp(1, &returnValue);
+		}
 
-	delete[] wideStringPath;
+		delete[] wideStringPath;
 #ifdef WINDOWS
-	delete[] wideStringLibraryPath;
+		delete[] wideStringLibraryPath;
 #endif
+	}
+	catch (const std::exception& ex)
+	{
+		error_return(ex.what())
```

Comment:
another approach where the R interpreter appears to stop processing if this return is used.